subsys/bluetooth/host/iso.c

```diff
@@ -982,89 +932,183 @@ static struct net_buf *hci_le_set_cig_params(struct bt_iso_create_param *param)
+static struct bt_iso_cig *get_free_cig(void)
 {
-	struct bt_conn *conn;
+	/* We can use the index in the `cigs` array as CIG ID */
+
+	for (int i = 0; i < ARRAY_SIZE(cigs); i++) {
+		if (!cigs[i].initialized) {
+			cigs[i].id = i;
+			return &cigs[i];
+		}
+	}
+
+	BT_DBG("Could not allocate any more CIGs");
+
+	return NULL;
+}
+
+static int cig_init_cis(struct bt_iso_cig *cig)
+{
+	for (int i = 0; i < cig->num_cis; i++) {
+		struct bt_iso_chan *cis = cig->cis[i];
+
+		if (cis == NULL) {
+			BT_DBG("CIS was NULL");
+			return -EINVAL;
+		}
+
+		if (cis->conn) {
+			BT_DBG("CIS conn was already allocated");
+			return -EALREADY;
+		}
+
+		cis->conn = iso_new();
+		if (cis->conn == NULL) {
+			BT_ERR("Unable to allocate CIS connection");
+			return -ENOMEM;
+		}
+
+		cis->conn->iso.cig_id = cig->id;
+		cis->conn->iso.is_bis = false;
+		cis->conn->iso.cis_id = i;
+
+		bt_iso_chan_add(cis->conn, cis);
+	}
+
+	return 0;
+}
+
+static void cleanup_cig(struct bt_iso_cig *cig)
+{
+	for (int i = 0; i < cig->num_cis; i++) {
+		struct bt_iso_chan *cis = cig->cis[i];
+
+		if (cis != NULL && cis->conn != NULL) {
+			bt_conn_unref(cis->conn);
+			cis->conn = NULL;
+		}
+	}
+
+	memset(cig, 0, sizeof(*cig));
+}
+
+int bt_iso_cig_create(const struct bt_iso_cig_create_param *param,
+		      struct bt_iso_cig **out_cig)
+{
+	int err;
 	struct net_buf *rsp;
+	struct bt_iso_cig *cig;
 	struct bt_hci_rp_le_set_cig_params *cig_rsp;
-	int i, err;
 
-	/* Check if controller is ISO capable */
+	CHECKIF(out_cig == NULL) {
+		BT_DBG("out_cig is NULL");
+		return -EINVAL;
+	}
+
+	*out_cig = NULL;
+
+	/* Check if controller is ISO capable as a central */
 	if (!BT_FEAT_LE_CIS_MASTER(bt_dev.le.features)) {
 		return -ENOTSUP;
 	}
 
-	if (!param->num_conns || param->num_conns > CONFIG_BT_ISO_MAX_CHAN) {
+	CHECKIF(param->cis_channels == NULL) {
+		BT_DBG("NULL CIS channels");
 		return -EINVAL;
 	}
 
-	/* Assign ISO connections to each LE connection */
-	for (i = 0; i < param->num_conns; i++) {
-		conn = param->conns[i];
+	CHECKIF(param->num_cis == 0) {
+		BT_DBG("Invalid number of CIS %u", param->num_cis);
+		return -EINVAL;
+	}
 
-		if (conn->type != BT_CONN_TYPE_LE) {
-			err = -EINVAL;
-			goto failed;
-		}
+	cig = get_free_cig();
 
-		conn = bt_conn_add_iso(conn);
-		if (!conn) {
-			err = -ENOMEM;
-			goto failed;
-		}
+	if (!cig) {
+		return -ENOMEM;
+	}
 
-		conn->iso.cig_id = param->id;
-		conn->iso.cis_id = bt_conn_index(conn);
+	cig->cis = param->cis_channels;
+	cig->num_cis = param->num_cis;
 
-		param->conns[i] = conn;
+	err = cig_init_cis(cig);
+	if (err) {
+		BT_DBG("Could not init CIS %d", err);
+		cleanup_cig(cig);
+		return err;
 	}
 
-	rsp = hci_le_set_cig_params(param);
-	if (!rsp) {
+	rsp = hci_le_set_cig_params(cig, param);
+	if (rsp == NULL) {
+		BT_WARN("Unexpected response to hci_le_set_cig_params");
 		err = -EIO;
-		goto failed;
+		cleanup_cig(cig);
+		return err;
 	}
 
 	cig_rsp = (void *)rsp->data;
```

Comment:
Why cast to void, and not to the correct type?

Reply:
Not part of this PR (and not my code either). 

The reason is, however, that you can always assign any pointer to a void pointer. I normally prefer to use the target struct type as my cast, but some people don't.

Reply:
I prefer casting to the target size.  But not part of this PR, and not very important to change.